frontend/gateway/container.go

```diff
@@ -71,122 +53,202 @@ func NewContainer(ctx context.Context, e executor.Executor, sm *session.Manager,
-				if err != nil {
-					return nil, stack.Enable(err)
-				}
-				ctr.rootFS = mountWithSession(ref, g)
-			}
+type OutputRef struct {
+	Ref        cache.Ref
+	MountIndex int
+}
 
-			// delete root mount from list, handled here
-			mnts = append(mnts[:i], mnts[i+1:]...)
-			break
-		}
-	}
+type MakeMutable func(m *opspb.Mount, ref cache.ImmutableRef) (cache.MutableRef, error)
 
-	if ctr.rootFS.Src == nil {
-		return nil, errors.Errorf("root mount required")
-	}
+func PrepareMounts(ctx context.Context, mm *mounts.MountManager, cm cache.Manager, g session.Group, mnts []*opspb.Mount, refs []*worker.WorkerRef, makeMutable MakeMutable) (p PreparedMounts, err error) {
+	// loop over all mounts, fill in mounts, root and outputs
+	for i, m := range mnts {
+		var (
+			mountable cache.Mountable
+			ref       cache.ImmutableRef
+		)
 
-	for _, m := range mnts {
-		var ref cache.ImmutableRef
-		var mountable cache.Mountable
-		if m.WorkerRef != nil {
-			ref = m.WorkerRef.ImmutableRef
-			mountable = ref
+		if m.Dest == opspb.RootMount && m.MountType != opspb.MountType_BIND {
+			return p, errors.Errorf("invalid mount type %s for %s", m.MountType.String(), m.Dest)
+		}
 
-			if !m.Readonly {
-				var err error
-				mountable, err = makeMutable(m.WorkerRef.Worker, ref)
-				if err != nil {
-					return nil, stack.Enable(err)
-				}
+		// if mount is based on input validate and load it
+		if m.Input != opspb.Empty {
+			if int(m.Input) > len(refs) {
+				return p, errors.Errorf("missing input %d", m.Input)
 			}
+			ref = refs[int(m.Input)].ImmutableRef
+			mountable = ref
 		}
+
 		switch m.MountType {
 		case opspb.MountType_BIND:
-			// nothing to do here
+			// if mount creates an output
+			if m.Output != opspb.SkipOutput {
+				// if it is readonly and not root then output is the input
+				if m.Readonly && ref != nil && m.Dest != opspb.RootMount {
+					p.OutputRefs = append(p.OutputRefs, OutputRef{
+						MountIndex: i,
+						Ref:        ref.Clone(),
+					})
+				} else {
+					// otherwise output and mount is the mutable child
+					active, err := makeMutable(m, ref)
+					if err != nil {
+						return p, err
+					}
+					mountable = active
+					p.OutputRefs = append(p.OutputRefs, OutputRef{
+						MountIndex: i,
+						Ref:        active,
+					})
+				}
+			} else if (!m.Readonly || ref == nil) && m.Dest != opspb.RootMount {
+				// this case is empty readonly scratch without output that is not really useful for anything but don't error
+				active, err := makeMutable(m, ref)
+				if err != nil {
+					return p, err
+				}
+				p.Actives = append(p.Actives, active)
+				mountable = active
+			}
+
 		case opspb.MountType_CACHE:
-			mRef, err := mm.MountableCache(ctx, toProtoMount(m), ref, g)
+			active, err := mm.MountableCache(ctx, m, ref, g)
 			if err != nil {
-				return nil, err
+				return p, err
+			}
+			mountable = active
+			p.Actives = append(p.Actives, active)
+			if m.Output != opspb.SkipOutput && ref != nil {
```

Comment:
I wonder if having output in here is even supported. Maybe just error?

Reply:
At least client protects against this https://github.com/moby/buildkit/blob/c70058095bd131bfbdca4d0796a2feb042bd1de8/client/llb/exec.go#L307

Reply:
Hmm I haven't changed this conditional, it's the same before refactoring here:
https://github.com/moby/buildkit/blob/9369d536788e44b8b991745a68bd9966e0239141/solver/llbsolver/ops/exec.go#L294

Reply:
Ok, yeah I missed that input is cloned to be output here. Still a weird case but no changes needed for this PR.